File: koller/bw/cpp/src/Koller1.cpp
/* Roman Szlachtun, 272330
 *
 * Impelementacja jednokanałowego algorytmu odejmowania tła w C++.
 *
 * Przyjmuje kolejne klatki nagrania, konwertuje ich do odcieni szarości
 * i zwraca ramki detekcji ruchu.
 */

#include <Koller1.h>

#include <algorithm>
#include <memory>
#include <numeric>
#include <string>

Koller1::Koller1(double deltaThreshold, double alpha1, double alpha2, double minAreaFraction)
    : deltaThreshold_(deltaThreshold),
      alpha1_(alpha1),
      alpha2_(alpha2),
      morphKernel_(cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(3, 3))),
      minAreaFraction_(minAreaFraction)
{
}

std::vector<BoundingBox> Koller1::predict(const cv::Mat &grayFrame) {
    std::vector<std::vector<cv::Point>> contours;
    std::vector<cv::Vec4i> hierarchy;
    kollerDiff(grayFrame);

    cv::findContours(diff_, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
    return filterContours(contours);
}

void Koller1::kollerDiff(const cv::Mat &grayFrame) {
    /*
     * Implementuje formułę selekcyjnej aktualizacji tła:
     * B_{t+1} = B_t + (alpha1*(1 - M_t) + alpha2*M_t) * D_t
     */
    if (background_.empty()) {
        throw std::runtime_error("Model tła nie jest zainicjalizowany.");
    }

    cv::Mat gray32f;
    grayFrame.convertTo(gray32f, CV_32F);

    cv::Mat Dt = gray32f - background_;
    cv::Mat absDt;
    cv::absdiff(gray32f, background_, absDt);

    // Binarna maska Mt odzwierciedla przekroczenie progu w Dt
    cv::Mat Mt_u8;
    cv::compare(absDt, deltaThreshold_, Mt_u8, cv::CMP_GE); // zakres 0 albo 255
    cv::Mat Mt;
    Mt_u8.convertTo(Mt, CV_32F, 1.0 / 255.0); // zakres 0.0 albo 1.0

    cv::Mat alpha_map = alpha1_ * (1.0f - Mt) + alpha2_ * Mt;
    background_ = background_ + alpha_map.mul(Dt);

    cv::min(background_, 255.0, background_);
    cv::max(background_, 0.0, background_);

    diff_ = Mt_u8;
    cv::morphologyEx(diff_, diff_, cv::MORPH_OPEN, morphKernel_, cv::Point(-1, -1), 1);
    cv::morphologyEx(diff_, diff_, cv::MORPH_CLOSE, morphKernel_, cv::Point(-1, -1), 2);
}

void Koller1::backgroundModelInit(const std::vector<fs::path> &frameList, double downscaleFactor) {
    std::vector<cv::Mat> images;

    for (const auto &p: frameList) {
        auto gray = readGrayscale(p.string(), downscaleFactor);
        gray.convertTo(gray, CV_32F);
        images.push_back(gray);
    }

    background_ = imagesMedian(images);
}

cv::Mat Koller1::drawDetectionBoxes(const fs::path &imgPath, const std::vector<BoundingBox> &df, double downscaleFactor) {
    auto image = cv::imread(imgPath.string(), cv::IMREAD_COLOR_BGR);
    for (const auto &b: df) {
        cv::rectangle(image, {static_cast<int>(downscaleFactor * b.x1), static_cast<int>(downscaleFactor * b.y1)}, {
                          static_cast<int>(downscaleFactor * b.x2), static_cast<int>(downscaleFactor * b.y2)
                      }, cv::Scalar(0, 0, 255), 4);
    }
    return image;
}

cv::Mat Koller1::getDiff() const {
    return diff_;
}


cv::Mat Koller1::imagesMedian(const std::vector<cv::Mat> &stack) {
    if (stack.empty()) {
        throw std::runtime_error("Nie można obliczyć mediany pustego stosu obrazów.");
    }

    const bool isEvenSizedStack = stack.size() % 2 == 0;
    const int rows = stack[0].rows;
    const int cols = stack[0].cols;

    for (const auto &image: stack) {
        if (image.type() != CV_32F || image.rows != rows || image.cols != cols) {
            throw std::runtime_error("Obrazy na stosie nie są spójne.");
        }
    }

    cv::Mat median(rows, cols, CV_32F);
    std::vector<float> valuesPerPixel;
    valuesPerPixel.reserve(stack.size());

    for (int y = 0; y < rows; ++y) {
        // tablica wskaźników na wiersz `y` dla każdego z obrazów na stosie
        auto rowYPointerForStackImages = std::make_unique<const float *[]>(stack.size());
        for (size_t i = 0; i < stack.size(); ++i) {
            // pobierz wskaźnik na wiersz `y` obrazu `i`
            rowYPointerForStackImages[i] = stack[i].ptr<float>(y);
        }

        auto *medianRowYPointer = median.ptr<float>(y);
        for (int x = 0; x < cols; ++x) {
            valuesPerPixel.clear();
            for (size_t i = 0; i < stack.size(); ++i) {
                // dodajemy wartość piksela (x, y) z obrazu `i` na stosie
                valuesPerPixel.push_back(rowYPointerForStackImages[i][x]);
            }

            if (isEvenSizedStack) {
                auto midIt = valuesPerPixel.begin() + static_cast<long>(valuesPerPixel.size() / 2);
                // przestawia elementy w valuesPerPixel tak, że *midIt jest elementem, który by znalazł się na środku posortowanej tablicy
                // ale nie sortuje całej tablicy i działa w czasie liniowym
                std::ranges::nth_element(valuesPerPixel.begin(), midIt, valuesPerPixel.end());

                auto midIt2 = valuesPerPixel.begin() + static_cast<long>(valuesPerPixel.size() / 2 - 1);
                std::ranges::nth_element(valuesPerPixel.begin(), midIt2, valuesPerPixel.end());
                medianRowYPointer[x] = 0.5f * (*midIt + *midIt2);
            } else {
                auto midIt = valuesPerPixel.begin() + static_cast<long>(valuesPerPixel.size() / 2);
                std::ranges::nth_element(valuesPerPixel.begin(), midIt, valuesPerPixel.end());
                medianRowYPointer[x] = *midIt;
            }
        }
    }
    return median;
}

std::vector<BoundingBox> Koller1::filterContours(const std::vector<std::vector<cv::Point> > &contours) const {
    const int frameHeight = background_.rows;
    const int frameWidth = background_.cols;
    const int minAreaPx = std::max(1, static_cast<int>(minAreaFraction_ * frameHeight * frameWidth));

    std::vector<BoundingBox> out;
    for (const auto &c: contours) {
        double area = cv::contourArea(c);
        if (area < static_cast<double>(minAreaPx)) continue;

        cv::Rect r = cv::boundingRect(c);
        int x1 = std::max(0, r.x);
        int y1 = std::max(0, r.y);
        int x2 = std::min(frameWidth, r.x + r.width);
        int y2 = std::min(frameHeight, r.y + r.height);

        if (x2 > x1 && y2 > y1) {
            out.emplace_back(x1, y1, x2, y2);
        }
    }
    return out;
}

cv::Mat Koller1::readGrayscale(const fs::path &imgPath, double downscaleFactor) {
    cv::Mat gray = cv::imread(imgPath.string(), cv::IMREAD_GRAYSCALE);

    if (gray.empty()) throw std::runtime_error("Obraz nie jest znaleziony: " + imgPath.string());
    cv::resize(gray, gray, cv::Size(), 1.0 / downscaleFactor, 1.0 / downscaleFactor, cv::INTER_AREA);
    gray.convertTo(gray, CV_8U);
    return gray;
}

File: koller/bw/py/Koller1.py
"""
Roman Szlachtun, 272330

Implementacja jednokanałowego algorytmu odejmowania tła w Python.

Przyjmuje kolejne klatki nagrania, konwertuje ich do odcieni szarości
i zwraca ramki detekcji ruchu.
"""

import time
from dataclasses import dataclass
from pathlib import Path
from typing import List, Sequence, Optional

import cv2
import numpy as np
from numpy import ndarray

from DirectoryAnchor import DATASET_DIR, PROJECT_ROOT_DIR

cv2.setUseOptimized(True)


@dataclass(frozen=True)
class BoundingBox:
    x1: int
    y1: int
    x2: int
    y2: int


class KollerBw:
    deltaThreshold_: int
    alpha1_: float
    alpha2_: float
    morphKernel_: cv2.typing.MatLike
    minAreaFraction_: float

    def __init__(
        self,
        deltaThreshold: int,
        alpha1: float,
        alpha2: float,
        minAreaFraction: float,
    ) -> None:
        self.diff_: Optional[np.ndarray] = None
        self.background_: Optional[np.ndarray] = None
        self.deltaThreshold_: int = deltaThreshold
        self.alpha1_: float = alpha1
        self.alpha2_: float = alpha2
        self.morphKernel_: cv2.Mat = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))
        self.minAreaFraction_: float = minAreaFraction

    def predict(self, grayFrame: np.ndarray) -> List[BoundingBox]:
        self.diff_ = self.kollerDiff(grayFrame)
        contours, _ = cv2.findContours(self.diff_, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

        return self.filterContours(contours)

    def kollerDiff(self, grayFrame: np.ndarray) -> np.ndarray:
        """
        Implementuje formułę selekcyjnej aktualizacji tła:
        B_{t+1} = B_t + (alpha1*(1 - M_t) + alpha2*M_t) * D_t
        """
        assert grayFrame.ndim == 2, "Są dozwolone tylko obrazy o jednym kanale kolorowym."
        assert self.background_ is not None, "Model tła nie jest zainicjalizowany."

        Dt = grayFrame - self.background_
        # Binarna maska Mt odzwierciedla przekroczenie progu w Dt
        Mt = (np.abs(Dt) >= self.deltaThreshold_).astype(np.float32)

        alpha_map = self.alpha1_ * (1.0 - Mt) + self.alpha2_ * Mt
        self.background_ = self.background_ + alpha_map * Dt
        np.clip(self.background_, 0.0, 255.0, out=self.background_)

        diff = (Mt * 255.0).astype(np.uint8)
        diff = cv2.morphologyEx(diff, cv2.MORPH_OPEN, self.morphKernel_, iterations=1)
        diff = cv2.morphologyEx(diff, cv2.MORPH_CLOSE, self.morphKernel_, iterations=2)

        return diff

    def backgroundModelInit(self, frameList: List[Path], downscaleFactor: float) -> None:
        images: List[np.ndarray] = []

        for framePath in frameList:
            image = readGrayscale(framePath, downscaleFactor)
            images.append(image.astype(np.float32))

        self.background_ = np.median(images, axis=0).astype(np.float32)

    def filterContours(self, contours: Sequence[ndarray]) -> List[BoundingBox]:
        """
        Przekształca kontury obszarów ruchu w prostokątne ramki
        """
        frameHeight, frameWidth = self.background_.shape
        minAreaPx = max(1, int(self.minAreaFraction_ * frameHeight * frameWidth))
        boxes: List[BoundingBox] = []

        for contour in contours:
            area = cv2.contourArea(contour)
            if area < minAreaPx:
                continue

            x, y, w, h = cv2.boundingRect(contour)
            x1 = int(max(0, x))
            y1 = int(max(0, y))
            x2 = int(min(frameWidth, x + w))
            y2 = int(min(frameHeight, y + h))

            if x2 > x1 and y2 > y1:
                boxes.append(BoundingBox(x1, y1, x2, y2))

        return boxes

    def drawDetectionBoxes(self, imgPath: Path, boxes: List[BoundingBox], downscaleFactor: float) -> np.ndarray:
        """
        Pozycje ramek detekcji są mnożone o downscaleFactor i
        nakładane na wejściową klatkę
        """
        img = cv2.imread(str(imgPath), cv2.IMREAD_COLOR_BGR)
        for box in boxes:
            cv2.rectangle(
                img,
                (int(box.x1 * downscaleFactor), int(box.y1 * downscaleFactor)),
                (int(box.x2 * downscaleFactor), int(box.y2 * downscaleFactor)),
                (0, 0, 255),
                thickness=2,
            )

        return img


def readGrayscale(imgPath: Path, downscaleFactor: float) -> np.ndarray:
    img = cv2.imread(str(imgPath), cv2.IMREAD_GRAYSCALE)
    img = cv2.resize(
        img,
        (int(img.shape[1] / downscaleFactor), int(img.shape[0] / downscaleFactor)),
        interpolation=cv2.INTER_AREA,
    )
    return img


def main():
    startImg = 4050
    endImg = 4500
    downscaleFactor = 1.0
    imagesCountForBackground = 20
    backgroundInitImgStep = 25

    saveResult = True
    saveMask = True

    detector = KollerBw(
        deltaThreshold=15,
        alpha1=2e-3,
        alpha2=2e-4,
        minAreaFraction=15e-4,
    )

    assert downscaleFactor >= 1.0, "Współczynnik przeskalowywania w dół nie może być mniejszy od 1."
    assert imagesCountForBackground > 0
    assert 0 < startImg <= endImg <= 4500

    backgroundStartImg = (startImg - 1) - (imagesCountForBackground - 1) * backgroundInitImgStep
    backgroundEndImg = startImg - 1

    startTime = time.time()
    initImgs = []
    for i in range(backgroundStartImg, backgroundEndImg + 1, backgroundInitImgStep):
        imgPath = DATASET_DIR / "frames" / f"{i}.png"
        initImgs.append(imgPath)

    detector.backgroundModelInit(initImgs, downscaleFactor)
    print(f"Czas inicjalizacji modelu tła: {time.time() - startTime:.3f}s")

    total = 0.0
    for i in range(startImg, endImg + 1):
        imgPath = DATASET_DIR / "frames" / f"{i}.png"

        startTime = time.time()
        boxes = detector.predict(readGrayscale(imgPath, downscaleFactor=downscaleFactor))
        print(f"Czas detekcji obiektów na obrazie {imgPath.name}: {time.time() - startTime:.3f}s")
        total += time.time() - startTime

        if saveResult:
            result = detector.drawDetectionBoxes(imgPath, boxes, downscaleFactor=downscaleFactor)
            cv2.imwrite(str(PROJECT_ROOT_DIR / "benchmark" / "results" / "koller1" / f"{i}-image.jpg"), result)

        if saveMask:
            cv2.imwrite(str(PROJECT_ROOT_DIR / "benchmark" / "results" / "koller1" / f"{i}-mask.jpg"), detector.diff_)

    if total != 0.0:
        print(f"Średni czas detekcji na klatkę: {total / (endImg - startImg + 1)}")
        print(f"Średni FPS: {1.0 / (total / (endImg - startImg + 1))}")


if __name__ == "__main__":
    main()
